dusty/source.py

```diff
@@ -52,32 +23,93 @@ def git_error_handling():
-        os.makedirs(repo_path_parent)
-    with git_error_handling():
-        repo = git.Repo.clone_from('ssh://{}@{}'.format(constants.GIT_USER, repo_name), repo_path)
-
-def update_local_repo(repo_name):
-    """Given a repo name (e.g. github.com/gamechanger/gclib), pull the latest
-    commits from master to bring the local copy up to date."""
-    ensure_local_repo(repo_name)
-
-    logging.info('Updating local repo {}'.format(repo_name))
-    notify('Pulling latest updates for {}'.format(short_repo_name(repo_name)))
-
-    repo_path = managed_repo_path(repo_name)
-    repo = git.Repo(repo_path)
-    with git_error_handling():
-        repo.remote().pull('master')
+class Repo(object):
+    def __init__(self, remote_path):
+        # remote path is either of actual remote (github.com/gamechanger/dusty) or
+        # other local repo (/repos/dusty) format
+        self.remote_path = remote_path
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return self.remote_path == other.remote_path
+        return False
+
+    def __hash__(self):
+        return hash(self.remote_path)
+
+    @classmethod
+    def resolve(cls, all_known_repos, name):
+        """We require the list of all remote repo paths to be passed in
+        to this because otherwise we would need to import the spec assembler
+        in this module, which would give us circular imports."""
+        match = None
+        for repo in all_known_repos:
+            if repo.remote_path == name: # user passed in a full name
+                return repo
+            if name == repo.short_name:
+                if match is None:
+                    match = repo
+                else:
+                    raise RuntimeError('Short repo name {} is ambiguous. It matches both {} and {}'.format(name,
+                                                                                                           match.remote_path,
+                                                                                                           repo.remote_path))
+        if match is None:
+            raise RuntimeError('Short repo name {} does not match any known repos'.format(name))
+        return match
+
+    @property
+    def is_local_repo(self):
+        return self.remote_path.startswith('/')
+
+    @property
+    def short_name(self):
+        return self.remote_path.split('/')[-1]
+
+    @property
+    def managed_path(self):
+        return os.path.join(constants.REPOS_DIR, self.remote_path.lstrip('/'))
+
+    @property
+    def is_overridden(self):
+        return self.remote_path in get_config_value(constants.CONFIG_REPO_OVERRIDES_KEY)
+
+    @property
+    def override_path(self):
+        repo_overrides = get_config_value(constants.CONFIG_REPO_OVERRIDES_KEY)
+        return repo_overrides.get(self.remote_path)
+
+    @property
+    def local_path(self):
+        return self.override_path if self.is_overridden else self.managed_path
```

Comment:
it never means the remote path, it is either the local path of your override directory or the local path of the managed directory